src/layout/dom.ts

```diff
@@ -59,23 +78,29 @@ export function getId(node: Node, autogen: boolean = false): number {
     return id ? id : null;
 }
 
-export function add(node: Node, data: NodeInfo, source: Source): void {
+export function add(node: Node, parent: Node, data: NodeInfo, source: Source): void {
     let id = getId(node, true);
-    let parentId = node.parentElement ? getId(node.parentElement) : null;
+    let element = node as HTMLElement;
+    let parentId = parent ? getId(parent) : null;
     let nextId = getNextId(node);
     let masked = true;
-    let parent = null;
+    let parentValue = null;
     let region = regionMap.has(node) ? regionMap.get(node) : null;
 
     if (parentId >= 0 && values[parentId]) {
-        parent = values[parentId];
-        parent.children.push(id);
-        region = region === null ? parent.region : region;
-        masked = parent.metadata.masked;
+        parentValue = values[parentId];
+        parentValue.children.push(id);
+        region = region === null ? parentValue.region : region;
+        masked = parentValue.metadata.masked;
     }
 
-    if (data.attributes && Constant.MASK_ATTRIBUTE in data.attributes) { masked = true; }
-    if (data.attributes && Constant.UNMASK_ATTRIBUTE in data.attributes) { masked = false; }
+    // If element has a valid shadowRoot, track Shadow DOM as a top level root node.
```

Comment:
What happens if the shadow DOM is attached in closed rather than open mode? In that case shadowRoot will be null. Will that affect our ability to playback the page? Should we check for that case here explicitly and send back an indication that there was a shadom that was attached in closed mode and, therefore, something could not be captured?

Reply:
If shadow dom is closed then yes we can't access (similar to third party IFRAME). Should we monkey patch? Not in favor - at least not yet. I think it's ok for Clarity to proxy the calls but not alter the calls. For instance if a web component really wanted a closed shadow dom - we should continue with that and not break the page (unintentionally by changing it to open). Also, once this problem happens in practice, we can revisit it. I have not seen any real use case for open shadow dom or encountered one yet.